Paint in React - Cristian Bendea

ChrisTheSpud opened the pull request:
- UI needs some more work;
- Missing Shape Fill and Help;
- Undo/redo are wonky in some cases;
- Code can use some refactoring, if possible;

ChrisTheSpud committed:
First Commit

ChrisTheSpud committed:
Update to drawing functions

Removed duplicate code.
Mouse drawing broken as of this commit.

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,357 @@
+      <div className="container">
+        <div className='toolbar'>
+          <Undo
+            undoHandler={undo}></Undo>
```
Folosește Self-Closing-Tag când nu ai conținut. Sunt mai multe cazuri de genul.

```suggestion
          <Undo undoHandler={undo} />
```

ChrisTheSpud replied:
Am pus Self-Closing-Tag unde am considerat ca ar arata mai ok, dar probabil o sa mai modific cand incerc sa mai refac componentele.

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,357 @@
+  });
+
+  window.addEventListener("mouseup", endDrawing);
+  window.addEventListener("touchend", endDrawing);
```
Event-urile pe `window` ar trebui puse în efecte.

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,357 @@
+
+    } else {
+      
+      const drawCanvas = drawCanvasRef.current;
+      const context = drawCanvas.getContext('2d');
+      
+      let position;
+
+      if (isTouch) {
+        position = getTouchPos(drawCanvas, e);
+      } else {
+        position = getMousePos(drawCanvas, e);
+      }
+  
+      let { x , y } = position;
```
Ai multe linii care se repetă la fiecare if. Ai putea să le pui deasupra primului, ca să mai scurtezi codul.

ChrisTheSpud replied:
Am mutat toata logica pentru isTouch la inceput.

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,357 @@
+      
+      let { x , y } = position;
+
+      context.clearRect(0, 0, canvas.width, canvas.height);
+
+      const width = x - rectStart.x;
+      const height = y - rectStart.y;
+  
+      const centerX = rectStart.x + width / 2;
+      const centerY = rectStart.y + height / 2;
+      const radiusX = Math.abs(width / 2);
+      const radiusY = Math.abs(height / 2);
+  
+      context.beginPath();
+      context.ellipse(centerX, centerY, radiusX, radiusY, 0, 0, 2 * Math.PI);
+      context.stroke();
```
Eu aș face funcții dedicate pentru fiecare formă (i.e. `drawSquare`, `drawEllipse`, etc.). Funcția `draw` ar apela acele funcții.

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,357 @@
+
+    if (square.checked) {
+      const drawCanvas = drawCanvasRef.current;
+      const context = drawCanvas.getContext('2d');
```
Aș pune contextul separat, pentru a nu îl cere mereu. Poate cu un `useMemo` ([docs](https://react.dev/reference/react/useMemo))

```js
const context = useMemo(() => {
    const drawCanvas = drawCanvasRef.current;
    
    if(! drawCanvas) return null;
    return drawCanvas.getContext('2d');
}, [drawCanvasRef.current])
```

ChrisTheSpud replied:
Am facut ceva asemanator cu "useCallback".

vladutcornel reviewed (commented):
**Folosește Self-Closing-Tag.**

ChrisTheSpud committed:
Update App.jsx

Added "useCallback" for handling getting the contexts;
Added functions for each shape that is drawn;
Added "useEffect" for window events;
Shortened the draw function.

ChrisTheSpud committed:
Introduced Context

ChrisTheSpud committed:
Created a Canvas and Toolbar component

ChrisTheSpud committed:
Update to SCSS and Context

ChrisTheSpud committed:
Revert "Update to SCSS and Context"

This reverts commit d8d5a0bf27aeca3ee2b3530157c890a9b47f1b6d.

ChrisTheSpud committed:
Update to Components, SCSS, Context

ChrisTheSpud committed:
Added Shape Fill

vladutcornel commented on src/Components/DrawingCanvas.jsx:
```diff
@@ -0,0 +1,308 @@
+
+        if(document.getElementById('solid-fill').checked) {
+          context.fillRect(rectStart.x, rectStart.y, width, height);
+        };
```
Nu e nevoie să pui `;` dupa blocul de la `if`

vladutcornel commented on src/Components/DrawingCanvas.jsx:
```diff
@@ -0,0 +1,308 @@
+        } else if (ellipse.checked) {
+          drawEllipse(x,y);
+    
+        } else if (triangle.checked) {
```
`square`, `ellipse` și `triangle` nu sunt definite aici. Poate ai copiat codul de altundeva?

ChrisTheSpud replied:
Sunt id-urile de la forme, au ramas asa de cand aveam toate in acelasi fisier si nu am vazut sa imi arunce erori, asa ca am lasat asa. Am modificat in "getElementbyId".

vladutcornel commented on src/App.jsx:
```diff
@@ -0,0 +1,14 @@
+function App() {
+
+  return (
+    <>
```
Dacă ai un singur element, e redundant să pui fragment.

vladutcornel commented on src/Components/DrawingCanvas.jsx:
```diff
@@ -0,0 +1,308 @@
+      })
+    
+      window.addEventListener("keydown", (ev) => {
+          if (ev.key === "y" && ev.ctrlKey) {
+              redo();
+          } else if (ev.key === "y" && ev.metaKey) {
+              redo();
```
Ambele `if`-uri fac redo. Ai putea să combini condițiile într-una
```suggestion
          if (ev.key === "y" && (ev.ctrlKey || && ev.metaKey)) {
              redo();
```

la fel mai sus la `undo`

vladutcornel commented on src/Components/Toolbar.jsx:
```diff
@@ -0,0 +1,157 @@
+                />
+                <img 
+                    className='draw-image' 
+                    src={square}
```
La câteva imagini aici ai uitat să pui `alt`.

vladutcornel commented on src/Components/Toolbar.scss:
```diff
@@ -0,0 +1,98 @@
+    border: none;
+    cursor: pointer;
+    border-radius: 15px;
+    border: none;
```
Ai `border` de două ori.

vladutcornel commented on src/Contexts/DrawingContext.jsx:
```diff
@@ -0,0 +1,30 @@
+
+export const DrawingContext = createContext();
+
+export const canvasRef = () => {
```
Funcțiile care folosesc chestii de React (precum `useRef`) se numesc hooks și ar trebui să înceapă mereu cu `use`. Adică ar trebui să dai funcției numele `useCanvasRef`.

vladutcornel commented on src/Components/DrawingCanvas.jsx:
```diff
@@ -0,0 +1,308 @@
+
+    return (
+        <>
+          <DrawingContext.Provider value={{undo , redo , resetCanvas, strokeSize , eraser, colorPicker, fillCanvas}}>
```
Eu aș fi încercat să mut toate funcțiile astea într-un hook și aș da rezultatul la valoarea contextului. Ceva de genul:
```jsx
const canvas = useCanvasRef();
return <DrawingContext.Provider value={canvas}>
```

ChrisTheSpud committed:
Update dupa review
